### PR TITLE
Track: Track1; Team name: TopoCoop; Model: CoGNN

### PR DESCRIPTION
**Pull request by magoflaco**
## Model

This PR implements **Co-GNN (Cooperative Graph Neural Networks)** as a Track 1 (GNN) backbone for the TDL Challenge 2026.

> Ben Finkelshtein, Xingyue Huang, Michael Bronstein, İsmail İlkan Ceylan.
> **Cooperative Graph Neural Networks.** ICML 2024.
> Paper: https://arxiv.org/abs/2310.01267 — Reference code: https://github.com/benfinkelshtein/CoGNN

Co-GNN reframes message passing as a cooperative game: at every layer, each node chooses one of four actions — **Standard, Listen, Broadcast, Isolate** — predicted by lightweight *action networks* (Eq. (1) of the paper) and sampled with the straight-through Gumbel-softmax estimator. The *environment network* then propagates messages only along edges whose source broadcasts and whose target listens (Eq. (2)). This gives every node a learned, task-adaptive receptive field, which is the paper's central mechanism for handling **heterophily** and long-range information flow — precisely the structural axes the GraphUniverse evaluation sweeps.

## What is included

| Component | Path |
|---|---|
| Backbone (`CoGNN`, action nets, weighted convs, learnable temperature) | `topobench/nn/backbones/graph/cognn.py` |
| Config, Co-GNN(μ,μ) variant (default) | `configs/model/graph/cognn.yaml` |
| Config, Co-GNN(Σ,Σ) variant | `configs/model/graph/cognn_sum.yaml` |
| Unit tests (100% line coverage of the new module) | `test/nn/backbones/graph/test_cognn.py` |
| Pipeline test | `test/pipeline/test_pipeline.py` |
| Evaluation results (official notebook output) | `2026_tdl_challenge/results.json` |

The implementation follows the reference code (`models/CoGNN.py`, `models/action.py`, `models/layers.py`, `models/temp.py`) and the paper's equations; all docstrings reference the specific equations/sections they implement. The backbone reuses the standard `AllCellFeatureEncoder`, `GNNWrapper` and `NoReadOut` — no custom wrapper/encoder/readout is needed.

Implementation notes (deviations allowed under "Modifications to respect TopoBench computational requirements"):
- The reference model's input encoder is folded into the backbone as a linear layer; the task decoder/pooling is delegated to the TopoBench readout, matching the TBModel composition.
- The weighted convolutions are implemented with `torch_geometric.utils.scatter` rather than `MessagePassing` subclasses, because TopoBench's file-based backbone auto-discovery is incompatible with PyG's `MessagePassing` inspector (module not registered in `sys.modules`). The math is identical to the reference layers.
- The four-way action is factorized into two binary Gumbel-softmax decisions (listen / broadcast) exactly as in the reference implementation.

## Evaluation

`2026_tdl_challenge/run_evaluation.ipynb` was run on a Kaggle P100 GPU with `MODEL_CONFIG = "graph/cognn"` (72 runs: 12 grid settings × 3 seeds × 2 tasks, full OOD evaluation; 2.6 h wall time; 31,941 parameters, ≈1.5 s/epoch). The generated `results.json` is committed. Community-detection accuracy averages 0.63 under high homophily and — consistent with the paper's central claim about learned, task-adaptive information flow — remains at 0.33 under severe heterophily (h ∈ [0, 0.1], 20 classes), far above chance.

Note on the notebook integrity check: current `main` ships a stale `expected_hash` that fails on a clean clone (#330). To run the notebook we updated only the `expected_hash` cell (which is above the `# UNIQUE_HASH_MARKER`, i.e. in the same user-editable region as `MODEL_CONFIG`) to the SHA-256 of the untouched protected cells. All cells from the marker onward are byte-identical to `main`.

## Tests

- `test/nn/backbones/graph/test_cognn.py`: 12 tests covering the backbone (all four layer types, skip/layer-norm/learned-temperature options, gradient flow through the straight-through estimator, the Eq. (2) edge-gating rule, and the weighted convolutions), 100% line coverage of `cognn.py`.
- `test/pipeline/test_pipeline.py`: end-to-end training of both variants on `graph/MUTAG`.

## Result figures

In-distribution test performance across the 12 GraphUniverse grid settings (mean ± std over 3 seeds):

**Community detection (accuracy):**

Community detection accuracy heatmap

**Triangle counting (MSE / total triangles):**

Triangle counting MSE heatmap

<details>
<summary>OOD delta plots (OOD − in-distribution), by training homophily level</summary>

OOD low homophily, community detection
OOD mid homophily, community detection
OOD high homophily, community detection
OOD low homophily, triangle counting
OOD mid homophily, triangle counting
OOD high homophily, triangle counting

</details>

### CODE REVIEW

File: test/nn/backbones/graph/test_cognn.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        out = conv(x, edges_1)
+        assert out.shape == (x.shape[0], 8)
+        out = conv(x, edges_1, edge_weight=edge_weight)
+        assert out.shape == (x.shape[0], 8)
+
+    # Zero edge weights must block all messages: for WeightedGNNConv the
+    # aggregated neighbor part is zero, so outputs for isolated updates
+    # coincide with using a zeroed neighborhood.
+    conv = WeightedGNNConv(x.shape[1], 8, aggr="sum")
+    zero_w = torch.zeros(edges_1.shape[1])
```
[the remark this answers is not in the document]
Comment (magoflaco) [reply test/nn/backbones/graph/test_cognn.py:205]: Good suggestion - added `testWeightedAggregate_mean_gating` in cca67a2, which pins the exact mean semantics with mixed 0/1 gates (one active + one gated neighbor into the same node).

File: topobench/nn/backbones/graph/cognn.py [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+    -------
+    torch.Tensor
+        Aggregated messages of shape [num_nodes, channels].
+    """
+    messages = x[edge_index[0]]
+    if edge_weight is not None:
+        messages = edge_weight.view(-1, 1) * messages
+    return scatter(
+        messages, edge_index[1], dim=0, dim_size=num_nodes, reduce=aggr
+    )
```
[the remark this answers is not in the document]
Comment (magoflaco) [reply topobench/nn/backbones/graph/cognn.py:67]: This dilution is intentional and matches the official Co-GNN implementation: in the reference code (`models/layers.py`, `MolConv`), messages are scaled by `edge_weight` inside `message()` and then reduced with PyG's standard `aggr='mean'`, which divides by the full in-degree - including gated (weight-0) edges. That is the exact MeanGNN behavior used to produce the paper's reported results, so a 'true weighted mean' would deviate from the architecture as published. Documented explicitly in the `weighted_aggregate` docstring and pinned by a regression test in cca67a2.

File: test/pipeline/test_pipeline.py [hunk 3]
```diff
@@ -7,7 +7,7 @@
 
 
 DATASET = "graph/MUTAG"  # ADD YOUR DATASET HERE
-MODELS = ["graph/gcn", "cell/topotune", "simplicial/topotune"]  # ADD ONE OR SEVERAL MODELS
+MODELS = ["graph/cognn", "graph/cognn_sum"]  # ADD ONE OR SEVERAL MODELS
```
[the remark this answers is not in the document]
Comment (magoflaco) [reply test/pipeline/test_pipeline.py]: Fair point - restored the upstream model list (`graph/gcn`, `cell/topotune`, `simplicial/topotune`) alongside both CoGNN variants in cca67a2.
